Mathlib/Topology/UniformSpace/CompletelyRegular.lean

```diff
@@ -0,0 +1,325 @@
+      rw [Filter.mem_principal, idRel_subset]
+      exact fun x => refl_mem_uniformity <| uniformity_le_uniformSpaceOf s_in_unf r_in_unf
+    symm := by
+      rw [Filter.Tendsto, Filter.le_generate_iff]
+      intro r ⟨n, r_eq⟩
+      rw [← r_eq, Filter.mem_sets, Filter.mem_map, UniformSpace.uniformityBasisOf_symmetric]
+      apply Filter.mem_generate_of_mem
+      use n
+    comp := by
+      rw [Filter.le_generate_iff]
+      intro r ⟨n, r_eq⟩
+      rw [← r_eq]
+      apply Filter.mem_of_superset _ <| UniformSpace.uniformityBasisOf_comp_subset s_in_unf n
+      apply (Filter.mem_lift'_sets <| Monotone.compRel monotone_id monotone_id).mpr
+      exact ⟨_, Filter.mem_generate_of_mem ⟨n + 1, rfl⟩, subset_refl _⟩
+  }
+
+theorem self_mem_uniformSpaceOf (s_in_unf : s ∈ 𝓤 X) :
+    s ∈ (UniformSpaceOf s_in_unf).uniformity :=
+  Filter.mem_of_superset (Filter.mem_generate_of_mem (Exists.intro 0 rfl))
+    <| (symmetrizeRel_subset_self _)
+
+/--
+The ambient uniformity is equal to the infimum of all `UniformSpaceOf` for `s ∈ 𝓤 X`.
+-/
+theorem uniformity_eq_iInf_uniformSpaceOf :
+    𝓤 X = ⨅ s, ⨅ (h : s ∈ 𝓤 X), 𝓤[UniformSpaceOf h] :=
+  _root_.le_antisymm
+    (by
+      simp_rw [le_iInf_iff]
+      exact fun _ mem_unf => uniformity_le_uniformSpaceOf mem_unf)
+    (fun (r : Set (X × X)) (r_in_unf : r ∈ 𝓤 X) =>
+      Filter.mem_iInf_of_mem r
+        <| Filter.mem_iInf_of_mem r_in_unf
+        <| self_mem_uniformSpaceOf r_in_unf)
+
+/--
+The uniformity associated with `s ∈ 𝓤 X` is countably generated, so a pseudometric space
+can be constructed for it.
+
+Note that the uniformity associated with `s ∈ 𝓤 X` is in general
+not equal to the ambient uniformity, see `UniformSpace.UniformSpaceOf` for more details.
+-/
+noncomputable def PseudoMetricSpaceOf (s_in_unf : s ∈ 𝓤 X) : PseudoMetricSpace X :=
+  @UniformSpace.pseudoMetricSpace X (UniformSpaceOf s_in_unf)
+    ⟨⟨_, Set.countable_range fun n => (UniformityBasisOf s_in_unf n).val, rfl⟩⟩
+
+/--
+If a set `t` is open within the topology generated by `UniformSpaceOf s_in_unf`,
+then one can find some `ε > 0` for all of its points
+-/
+theorem isOpen_uniformSpaceOf_iff (s_in_unf : s ∈ 𝓤 X) {t : Set X} :
+    IsOpen[(UniformSpaceOf s_in_unf).toTopologicalSpace] t ↔
+      ∀ x ∈ t, ∃ ε, 0 < ε ∧ @Metric.ball X (PseudoMetricSpaceOf s_in_unf) x ε ⊆ t :=
+  @Metric.isOpen_iff X (PseudoMetricSpaceOf s_in_unf) _
+
+/--
+If `x ∈ s` for some open `s`, then there exists a family of pseudometric spaces `D`
+and an `ε > 0`, such that:
+* `∀ m ∈ D`, the topology associated with `D` is coarser than the ambient topology on `X`.
+* The intersection of the balls $$B_d(x, ε)$$ for all `d ∈ D` is a subset of `s`.
+-/
+theorem exists_pseudoMetricSpace_family_of_isOpen {s : Set X} (s_open : IsOpen s)
+    {x : X} (x_in_s : x ∈ s) : ∃ D : Set (PseudoMetricSpace X), D.Finite ∧
+      (∀ m ∈ D, unf.toTopologicalSpace ≤ m.toUniformSpace.toTopologicalSpace) ∧
+      ∃ (ε : ℝ), 0 < ε ∧ ⋂ m ∈ D, @Metric.ball X m x ε ⊆ s := by
```

Comment:
I think in this proof you're making your life harder by trying to use the explicit family of metrizable uniformities that you've built. 

What I would do is start by this general lemma, which doesn't mention `uniformSpaceOf` explicitly:
```lean
variable (X) in
theorem UniformSpace.eq_iInf_pseudoMetric :
    unf = ⨅ (m : PseudoMetricSpace X) (_ : unf ≤ m.toUniformSpace), m.toUniformSpace :=
  sorry
```

Then, you can do something like this:
```lean
theorem exists_pseudoMetricSpace_family_of_nhds {s : Set X} {x : X} (hxs : s ∈ 𝓝 x) :
    ∃ D : Set (PseudoMetricSpace X), D.Finite ∧
      (∀ m ∈ D, unf.toTopologicalSpace ≤ m.toUniformSpace.toTopologicalSpace) ∧
      ∃ (ε : ℝ), 0 < ε ∧ ⋂ m ∈ D, @Metric.ball X m x ε ⊆ s := by
  rw [UniformSpace.eq_iInf_pseudoMetric X, iInf_subtype'] at hxs
  simp_rw [toTopologicalSpace_iInf, nhds_iInf, Filter.mem_iInf'] at hxs
  rcases hxs with ⟨D, D_finite, V, hVx, -, rfl, -⟩
  use (↑) '' D, D_finite.image _, ball_image_of_ball fun m hm ↦ toTopologicalSpace_mono m.2
  simp_rw [@Metric.mem_nhds_iff X (_)] at hVx
  choose ε ε_pos hεV using hVx
  by_cases D_nonempty : D.Nonempty
  · refine ⟨sInf (ε '' D), ?_, ?_⟩
    · rw [((D_finite.image _).lt_cInf_iff (D_nonempty.image _)), ball_image_iff]
      exact fun m _ ↦ ε_pos m
    · rw [biInter_image]
      refine iInter₂_mono fun m hm ↦ (@Metric.ball_subset_ball X (_) _ _ _ ?_).trans (hεV m)
      exact csInf_le (D_finite.image _).bddBelow (mem_image_of_mem _ hm)
  sorry -- stupid case...
```

Of course this second proof could very easily use `uniformity_eq_iInf_uniformSpaceOf` directly, but I think it's cleaner to indicate that it really doesn't rely on the specific form of the uniformities at play.

Reply:
In fact, what do you think of hiding the full construction inside `UniformSpace.eq_iInf_pseudoMetric` (or some version where you take the infimum on countably generated uniformities)? I don't think one ever needs to use the explicit family `UniformSpaceOf`, do you have any example in mind where it's useful?

Reply:
Two extra things:
- I assumed `s ∈ 𝓝 x` because it's enough, and it highlights that this is really a local lemma
- Please keep your comments! More generally, feel free to make the proof longer for readability, my point was to illustrate how to cut some arguments, not to golf the proof.